replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -13,140 +23,96 @@ private[journal] sealed abstract class Batch extends Product {
 private[journal] object Batch {
 
   def of(records: NonEmptyList[ActionRecord[Action]]): List[Batch] = {
+    records.reverse.foldLeft(State()) { _.handle(_) }.batches
+  }
 
-    // returns `true` when we can optimize Cassandra usage by NOT inserting "previous" Append actions in DB
-    // because "current" Delete action will discard them all
-    def cut(appends: Appends, delete: Action.Delete): Boolean = {
-      val append = appends.records.head.action
-      append.range.to <= delete.to.value
-    }
+  private final case class State(
+    purge: Option[Purge]     = None,
+    appends: Option[Appends] = None,
+    delete: Option[Delete]   = None,
+  ) {
 
-    def dropDeleted(appends: Appends, delete: Action.Delete): Option[Appends] = {
-      // we cannot drop individual events in record as we do not parse the payload here (`records.head.action.payload: ByteVector`)
-      val validRecords = appends.records.collect { case record if record.action.range.to >= delete.to.value => record }
-      NonEmptyList.fromList(validRecords).map { Appends(appends.offset, _) }
-    }
+    def handle: ActionRecord[Action] => State = {
```

Comment:
As this `handle` method relies on the records being handled in reverse order, it might be worth to name it accordingly or just mention it in the comments.